File: Editorial/0123_Best_Time_to_Buy_and_Sell_Stock_III/Go/1_Bidirectional_Dynamic_Programming.go
func max(x int, y int) int {
    if x > y {
        return x
    }
    return y
}

func min(x int, y int) int {
    if x > y {
        return y
    }
    return x
}

func maxProfit(prices []int) int {
    length := len(prices)
    if length <= 1 {
        return 0
    }
    leftMin := prices[0]
    rightMax := prices[length-1]
    leftProfits := make([]int, length)
    rightProfits := make([]int, length+1)
    for l := 1; l < length; l++ {
        leftProfits[l] = max(leftProfits[l-1], prices[l]-leftMin)
        leftMin = min(leftMin, prices[l])
        r := length - 1 - l
        rightProfits[r] = max(rightProfits[r+1], rightMax-prices[r])
        rightMax = max(rightMax, prices[r])
    }
    maxProfit := 0
    for i := 0; i < length; i++ {
        maxProfit = max(maxProfit, leftProfits[i]+rightProfits[i+1])
    }
    return maxProfit
}

File: Editorial/0101-0200/0123_Best_Time_to_Buy_and_Sell_Stock_III/C/1_Bidirectional_Dynamic_Programming.c
int max(int x, int y) { return (x > y) ? x : y; }
int min(int x, int y) { return (x > y) ? y : x; }
int maxProfit(int* prices, int pricesSize) {
    if (pricesSize <= 1) return 0;
    int leftMin = prices[0];
    int rightMax = prices[pricesSize - 1];
    int leftProfits[pricesSize];
    int rightProfits[pricesSize + 1];
    memset(rightProfits, 0, sizeof(rightProfits));
    int base = 0;
    memset(leftProfits, base, sizeof(leftProfits));
    for (int l = 1; l < pricesSize; ++l) {
        leftProfits[l] = max(leftProfits[l - 1], prices[l] - leftMin);
        leftMin = min(leftMin, prices[l]);
        int r = pricesSize - 1 - l;
        rightProfits[r] = max(rightProfits[r + 1], rightMax - prices[r]);
        rightMax = max(rightMax, prices[r]);
    }
    int maxProfit = 0;
    for (int i = 0; i < pricesSize; ++i)
        maxProfit = max(maxProfit, leftProfits[i] + rightProfits[i + 1]);
    return maxProfit;
}
